### PR TITLE
Make scoreboard not insane (WIP)

### PR DESCRIPTION
**Pull request by ObaniGemini**
Don't merge this, it's just for easier readability

### CODE REVIEW

File: base/progs/gametypes/generic/matchstates.as [hunk 1]
```diff
@@ -223,18 +175,3 @@ Entity @RandomEntity( String &className )
 		return null;
 	return spawnents[ random_uniform( 0, spawnents.size() ) ];
 }
-
-void GENERIC_UpdateMatchScore()
-{
-	if( gametype.isTeamBased ) {
```
Comment (mikejsavage) [inline base/progs/gametypes/generic/matchstates.as:229]: undecided between keeping this or letting the engine do it

for now let's keep it

File: source/cgame/cg_scoreboard.cpp [hunk 2]
```diff
@@ -4,78 +4,26 @@
...
 	const Material * icon = NULL;
 
 	if( warmup ) {
-		icon = player.state != 0 ? cgs.media.shaderReady : NULL;
+		icon = player->state ? cgs.media.shaderReady : NULL;
 	}
 	else {
-		bool carrier = player.state != 0 && ( ISREALSPECTATOR() || cg_entities[ id + 1 ].current.team == cg.predictedPlayerState.team );
-		if( alive ) {
+		bool carrier = player->state && ( ISREALSPECTATOR() || cg_entities[ playerIndice ].current.team == cg.predictedPlayerState.team );
```
Comment (mikejsavage) [inline source/cgame/cg_scoreboard.cpp]: index

File: source/cgame/cg_scoreboard.cpp [hunk 3]
```diff
@@ -288,9 +226,7 @@ void CG_DrawScoreboard() {
 		}
 	}
 	else {
-		ScoreboardTeam team_info;
-		if( !ParseTeam( &cursor, &team_info ) )
-			return;
+		SyncTeamState * team_info = &client_gs.gameState.teams[ TEAM_PLAYERS ];
```
Comment (mikejsavage) [inline source/cgame/cg_scoreboard.cpp]: can be const

File: source/game/g_ascript.cpp [hunk 4]
```diff
@@ -656,18 +652,23 @@ static const asClassDescriptor_t asTeamListClassDescriptor =
...
-static void objectScoreStats_ScoreSet( int newscore, score_stats_t *obj ) {
-	obj->score = newscore;
+static void objectScoreStats_SetScore( int score, score_stats_t *obj ) {
+	obj->score = score;
 }
 
-static void objectScoreStats_ScoreAdd( int score, score_stats_t *obj ) {
-	obj->score += score;
+static void objectScoreStats_SetState( bool state, score_stats_t * obj ) {
+	obj->state = state;
```
Comment (mikejsavage) [inline source/game/g_ascript.cpp]: Can you rename state to has_bomb or something

File: source/game/g_ascript.cpp [hunk 5]
```diff
@@ -1484,12 +1465,12 @@ static gclient_t *asFunc_GetClient( int clientNum ) {
 	return &game.clients[ clientNum ];
 }
 
-static g_teamlist_t *asFunc_GetTeamlist( int teamNum ) {
+static SyncTeamState * asFunc_GetTeamlist( int teamNum ) {
 	if( teamNum < TEAM_SPECTATOR || teamNum >= GS_MAX_TEAMS ) {
```
Comment (mikejsavage) [inline source/game/g_ascript.cpp]: should probably just be an assert since nobody checks the return value of this

File: source/game/g_frame.cpp [hunk 6]
```diff
@@ -151,6 +151,18 @@ static void G_UpdateServerInfo() {
 	}
 }
 
+static void G_UpdateClientScoreboard( edict_t * ent ) {
+	score_stats_t * stats = G_ClientGetStats( ent );
+	int plyNum = PLAYERNUM( ent );
```
Comment (mikejsavage) [inline source/game/g_frame.cpp]: could make server_gs.gamestate.players[i] a variable, also `stats` can be const

File: source/game/g_gameteams.cpp [hunk 7]
```diff
@@ -91,63 +91,17 @@ void G_Teams_UpdateMembersList() {
...
-
-/*
-* G_Teams_TeamIsLocked
-*/
-bool G_Teams_TeamIsLocked( int team ) {
-	if( team > TEAM_SPECTATOR && team < GS_MAX_TEAMS ) {
-		return teamlist[team].locked;
-	} else {
-		return false;
+		qsort( GetTeam( team ).playerIndices, GetTeam( team ).numplayers, sizeof( GetTeam( team ).playerIndices[0] ), G_Teams_CompareMembers );
```
Comment (mikejsavage) [inline source/game/g_gameteams.cpp]: could replace with std::sort + lambda while you're here

File: source/game/g_local.h [hunk 8]
```diff
@@ -160,11 +159,31 @@ struct spawn_temp_t {
...
+	int total_damage_given;
+	int total_damage_received;
+};
+
 extern game_locals_t game;
 extern gs_state_t server_gs;
 extern level_locals_t level;
 extern spawn_temp_t st;
 
+#define GetTeam( team ) server_gs.gameState.teams[ team ]
```
Comment (mikejsavage) [inline source/game/g_local.h]: would be nice to not have this

File: source/game/g_local.h [hunk 9]
```diff
@@ -544,7 +559,10 @@ void ClientUserinfoChanged( edict_t *ent, char *userinfo );
 void G_Client_UpdateActivity( gclient_t *client );
 void G_Client_InactivityRemove( gclient_t *client );
 void G_ClientRespawn( edict_t *self, bool ghost );
+score_stats_t * G_ClientGetStats( edict_t * ent );
+score_stats_t * G_ClientGetStats( gclient_t * client );
 void G_ClientClearStats( edict_t *ent );
+void G_ClientClearStats( gclient_t * client );
```
Comment (mikejsavage) [inline source/game/g_local.h]: can we get rid of one of these?

File: source/game/g_ascript.cpp (hunk 4 above)
Comment (ObaniGemini) [reply source/game/g_ascript.cpp]: no because it includes ready

**Comment by mikejsavage** [comment]
merged with https://github.com/mikejsavage/forksow/commit/06239084ea1424644fa03e7dea8fd2425ead96cf sry for jacking credit

**Pull request closed without merging**